File: backend/controllers/clinics.js
import { db } from "../config/db.js";

// get all clinics
export const getAllClinics = async (req, res) => {
    try {
        const result = await db.query(
            "SELECT * FROM clinics ORDER BY name ASC"
        );
        res.status(200).json(result.rows);
    } catch (error) {
        console.error("Error fetching clinics:", error);
        res.status(500).json({ error: "Failed to fetch all clinics" });
    }
};

// add a clinic
export const addClinic = async (req, res) => {
    try {
        const { name, address, phonenumber, email, latitude, longitude, services, operating_hours } = req.body;

        if (!name || !address || !phonenumber || !email) {
            return res.status(400).json({ message: "Clinic name, address, contact and email are required." });
        }

        await db.query(
            `INSERT INTO clinics (name, address, phonenumber, email, latitude, longitude, services, operating_hours) 
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
            [name, address, phonenumber, email, latitude, longitude, services, operating_hours]
        );
        res.status(200).json({ message: "Clinic added successfully." });
    } catch (error) {
        console.error("Error adding clinic:", error);
        res.status(500).json({ error: "Failed to add clinic." });
    }
};

File: frontend/lib/models/clinics.dart
class Clinics {
  final int id;
  final String name;
  final String address;
  final String phonenumber;
  final String email;
  final double latitude;
  final double longitude;
  final String services;
  final Map<String, dynamic> operatingHours;

  // constructor - required fields
  Clinics({
    required this.id,
    required this.name,
    required this.address,
    required this.phonenumber,
    required this.email,
    required this.latitude,
    required this.longitude,
    required this.services,
    required this.operatingHours,
  });

  // create clinics object from JSON data
  factory Clinics.fromJson(Map<String, dynamic> json) {
    return Clinics(
      id: json['id'],
      name: json['name'],
      address: json['address'],
      phonenumber: json['phonenumber'],
      email: json['email'],
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      services: json['services'],
      operatingHours: json['operating_hours'] ?? {},
    );
  }

  // convert clinics object to json
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name, 
      'address': address,
      'phonenumber': phonenumber, 
      'email': email, 
      'latitude': latitude, 
      'longitude': longitude, 
      'services': services,
      'operating_hours': operatingHours,
    };
  }
}